challenges/backend-challenge/src/main/java/com/number-submission/service/NumberSubmissionServiceImpl.java

```diff
@@ -0,0 +1,43 @@
+package com.numbersubmission.service;
+
+import java.util.List;
+
+@Service("numberSubmissionService")
+public class NumberSubmissionServiceImpl implements NumberService{
+		
+	private static List<Int> numberList;
+
+	public void saveNumberList(List<Int> numberList) {
+		this.numberList = numberList;
+	}
+
+	public List<Int> getLatestList() {
+		return numberList;
+	}
+	public List<Int> getSortedLatestList() {
+		if(!numberList.isEmpty()) {
+			return Collections.sort(numberList);
+		}
+		return new List<Int>();
+	}
+
+	public void insertValIntoList(Int number) {
+		if(!numberList.isEmpty()) {
+			List<Int> sortedNumberList = Collections.sort(numberList);
+			int index = 0;
+			// Find earliest sorted position or final index
+			while (sortedNumberList[index] < number || index == sortedNumberList.length() - 1) {
+				index++;
+			}
+			// Store val to replace
+			int valToReplace = sortedNumberList[index];
+			int index = 0;
+			while (numberList[index] != valToReplace) {
+				index ++;
+			}
+			// Replace at first occurence in original list
+			// (to avoid overlap w future implementations)
+			numberList[index] = number;
```

Comment:
Hi Nick, this was actually meant to keep the ordering of the original array the same and not to do sort of a shortcut and sort all of the values in the stored array and then remove the value from the array that gets replaced. That way, hypothetical future implementations could work with the original array for modification, printing, etc.

You do bring up a good point that concurrency might be an issue here. It might be a good idea to place locks outside of lines 28-40 to prevent overlap of the array assignment. Thanks for noticing this!

Cole

Reply:
Locks would be 1 way to solve this issue 👍 

There's another way to solve the concurrency issue using a functional approach. Are you familiar with functional programming/data immutability?

Reply:
If, say a second thread ran request 3 when thread 1(request 2) is on line 35, the index would go out of bounds i.e. would not find the val to replace in the array.

I'd need to add an index out of bounds catch when the patch inserts a val that does not exist in the array. I would also add a lock between lines 26 and 40 so that the array does not get overwritten during the insertion process.

Thanks for that catch!

Reply:
Interestingly enough, I was thinking that storing an immutable instance of this object might be the way to go, but wouldn't think that would want to be the answer because of the extra storage it'd take up. :)

Reply:
I’m glad you brought up the topic of immutability, because I actually brushed this off as not a preferred option. Glad to see that I had some rationale to my thinking. 🙂